### PR TITLE
Convert Reviews component to TypeScript

### LINKED ISSUE #7864: Convert components/reviews to TypeScript
We need to convert the `reviews` component to TypeScript.

Folder path: https://github.com/woocommerce/woocommerce-blocks/tree/trunk/assets/js/base/components/reviews

### PR DESCRIPTION
**Pull request by tarunvijwani**
With this PR, we're converting the Reviews component to TypeScript.

Fixes #7864

### Changes in the PR:
1. Converted the review-list component to TypeScript.
2. Converted the review-list-item component to TypeScript.
3. Converted the review-sort-select component to TypeScript.

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Checkout this branch
2. Go to Editor and add the `All Reviews` block. 
3. Confirm block is loaded fine in the Editor. 
4. Publish the page. 
5. Confirm block is getting displayed on the front-end. 

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### CODE REVIEW

File: assets/js/base/components/reviews/index.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
-export { default as ReviewList } from './review-list';
-export { default as ReviewListItem } from './review-list-item';
-export { default as ReviewSortSelect } from './review-sort-select';
+export { default as ReviewList } from './review-list/index';
```
Comment (gigitux) [inline assets/js/base/components/reviews/index.js]: Is there a reason why did you change these imports? Probably VSCode did it automatically. I think that it is not necessary

File: assets/js/base/components/reviews/types.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
+ * External dependencies
+ */
+import { EmptyObjectType } from '@woocommerce/types';
+
+export type review =
```
Comment (gigitux) [inline assets/js/base/components/reviews/types.ts]:
```suggestion
export type Review =
```

It is a good practice that the types start with capital letters.

File: assets/js/base/components/reviews/types.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+				name: string;
+				sizes: string;
+				src: string;
+				srcset: string;
+			};
+			reviewer_avatar_urls: { [ size: string ]: string };
+			verified: boolean;
+			rating: number;
+	  }
+	| EmptyObjectType;
```
Comment (gigitux) [inline assets/js/base/components/reviews/types.ts:29]: Not sure about ` | EmptyObjectType`. Is there a particular reason why did you add?

File: assets/js/base/components/reviews/review-sort-select/index.tsx [hunk 4]
```diff
@@ -2,15 +2,25 @@
...
 /**
  * Internal dependencies
  */
 import './style.scss';
 
-const ReviewSortSelect = ( { onChange, readOnly, value } ) => {
+interface ReviewSortSelectProps {
+	onChange: ChangeEventHandler;
+	readOnly: boolean;
+	value: string;
```
Comment (gigitux) [inline assets/js/base/components/reviews/review-sort-select/index.tsx]: Is there a reason why you defined the `value` type in this way? According to the old propTypes declaration, `value` can be only: 'most-recent', 'highest-rating', and 'lowest-rating'.

In TS, you can define it in this way 
```suggestion
	value: 'most-recent' | 'highest-rating' | 'lowest-rating'.;
```

File: assets/js/base/components/reviews/review-sort-select/index.tsx [hunk 5]
```diff
@@ -2,15 +2,25 @@
...
+import { ChangeEventHandler } from 'react';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
 
-const ReviewSortSelect = ( { onChange, readOnly, value } ) => {
+interface ReviewSortSelectProps {
+	onChange: ChangeEventHandler;
```
Comment (gigitux) [inline assets/js/base/components/reviews/review-sort-select/index.tsx:14]: It seems that onChange can be optional. You can achieve it in TS in this way:
```suggestion
	onChange?: ChangeEventHandler;
```

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for working on this! Nice work! I added some comments.

File: assets/js/base/components/reviews/review-sort-select/index.tsx (hunk 4 above)
Comment (tarunvijwani) [reply assets/js/base/components/reviews/review-sort-select/index.tsx]: > Is there a reason why you defined the value type in this way? According to the old propTypes declaration, value can be only: 'most-recent', 'highest-rating', and 'lowest-rating'.

The reason I switched it to String is because I was getting the below error in assets/js/blocks/reviews/frontend-block.js file:

```
Type 'string' is not assignable to type '"most-recent" | "highest-rating" | "lowest-rating"'.ts(2322)

```

Maybe there is a better way to fix it?

File: assets/js/base/components/reviews/types.ts (hunk 3 above)
Comment (tarunvijwani) [reply assets/js/base/components/reviews/types.ts:29]: > Not sure about | EmptyObjectType. Is there a particular reason why did you add?

We have the possibility of an empty object [here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/components/reviews/review-list-item/index.js#L153). If I remove it, it will throw TypeScript errors. I'm happy to learn if there is any other way we can approach this issue. 🙂

File: assets/js/base/components/reviews/review-sort-select/index.tsx (hunk 5 above)
Comment (tarunvijwani) [reply assets/js/base/components/reviews/review-sort-select/index.tsx:14]: > It seems that onChange can be optional. You can achieve it in TS in this way:

I'm not sure, the onChange function is a required prop in SortSelect: assets/js/base/components/sort-select/index.tsx

I tried changing both of them to optional. It was creating more issues, so I rolled back to required for now. Is there any other way to make them optional?

File: assets/js/base/components/reviews/review-sort-select/index.tsx (hunk 4 above)
Comment (gigitux) [reply assets/js/base/components/reviews/review-sort-select/index.tsx]: On `assets/js/blocks/reviews/frontend-block.js`, you could update the JSDoc in this way:
image

Comment (tarunvijwani) [reply assets/js/base/components/reviews/review-sort-select/index.tsx]: Thank you for the suggestion! I have updated the code. 👍

**Review by gigitux — APPROVED** [review]
Thanks for addressing the feedback! Great work! LGTM!

EDIT: I noticed that there is an error on `assets/js/base/components/reviews/review-list/index.tsx`:

image

You need update `review` to `Review`. I'm pre-approving this PR!

**Review by gigitux — APPROVED** [review]

**Comment by tarunvijwani** [comment]
>You need update review to Review. I'm pre-approving this PR!

Thank you, I made the changes. 👍
